contracts/finance/Vesting.sol

```diff
@@ -0,0 +1,284 @@
+        require(
+            vesting_.vestingStartTime > 0,
+            "VestingWallet: cannot create vesting for zero time"
+        );
+        require(
+            vesting_.vestingAmount > 0,
+            "VestingWallet: cannot create vesting for zero amount"
+        );
+        require(
+            vesting_.beneficiary != address(0),
+            "VestingWallet: cannot create vesting for zero address"
+        );
+
+        Schedule storage _schedule = _schedules[vesting_.scheduleId];
+
+        require(
+            vesting_.vestingStartTime +
+                _schedule.scheduleData.durationInPeriods *
+                _schedule.scheduleData.secondsInPeriod >
+                block.timestamp,
+            "VestingWallet: cannot create vesting for past date"
+        );
+
+        _vestingId++;
+
+        _beneficiaryIds[vesting_.beneficiary].add(_vestingId);
+
+        _vestings[_vestingId] = vesting_;
+
+        emit VestingCreated(_vestingId, vesting_.beneficiary, vesting_.vestingToken);
+
+        return _vestingId;
+    }
+
+    function _withdrawFromVesting(
+        uint256 vestingId_
+    ) internal virtual returns (uint256 _amountToPay, address _token) {
+        VestingData storage _vesting = _vestings[vestingId_];
+
+        require(
+            msg.sender == _vesting.beneficiary,
+            "VestingWallet: only beneficiary can withdraw from his vesting"
+        );
+
+        _amountToPay = getWithdrawableAmount(vestingId_);
+        _token = _vesting.vestingToken;
+
+        _vesting.paidAmount += _amountToPay;
+
+        emit WithdrawnFromVesting(
+            vestingId_,
+            _vesting.beneficiary,
+            _vesting.vestingToken,
+            _amountToPay
+        );
+    }
+
+    function _vestingCalculation(
+        Schedule memory schedule_,
+        uint256 totalVestingAmount_,
+        uint256 vestingStartTime_,
+        uint256 timestamp_
+    ) internal view virtual returns (uint256 _vestedAmount) {
+        BaseSchedule memory baseData_ = schedule_.scheduleData;
+
+        if (vestingStartTime_ > timestamp_) return _vestedAmount;
+
+        uint256 elapsedPeriods_ = _calculateElapsedPeriods(
+            vestingStartTime_,
+            timestamp_,
+            baseData_.secondsInPeriod
+        );
+
+        if (elapsedPeriods_ <= baseData_.cliffInPeriods) {
+            return 0;
+        }
+
+        if (elapsedPeriods_ >= baseData_.durationInPeriods) {
+            return totalVestingAmount_;
+        }
+
+        uint256 elapsedPeriodsPercentage_ = (elapsedPeriods_ * PRECISION) /
+            baseData_.durationInPeriods;
+
+        _vestedAmount =
+            (_raiseToPower(elapsedPeriodsPercentage_, schedule_.exponent) *
+                (totalVestingAmount_)) /
+            _raiseToPower(PRECISION, schedule_.exponent);
+
+        return _vestedAmount.min(totalVestingAmount_);
+    }
+
+    function _calculateElapsedPeriods(
```

Comment:
Why did you decide to make this private?

Reply:
I don't see how or why the user can edit or extend this function. The only point I see is that the user can use these functions for some calculations in their contract. If so, I can just make this function internal without making it virtual. Continuing with this, should we also make the '_raiseToPower' function internal, as it can also be used for some calculations?

Reply:
Let's make it internal without virtual